CMakeLists.txt

```diff
@@ -1,73 +1,71 @@
-    cmake_policy(SET CMP0074 NEW)
-endif()
+set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules)
 
 # TODO: fix the macOS bundle stuff
-if (POLICY CMP0080)
-    cmake_policy(SET CMP0080 OLD)
+cmake_policy(SET CMP0080 OLD)
+
+# this must be defined before project() or enable_language() to take effect
+if (WIN32)
+    if (GENERATOR_IS_MULTI_CONFIG)
+        set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<BOOL:${QBT_MSVC_RUNTIME_DYNAMIC}>:DLL>$<$<CONFIG:Debug>:Debug>")
+    else()
+        set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<BOOL:${QBT_MSVC_RUNTIME_DYNAMIC}>:DLL>$<$<CMAKE_BUILD_TYPE:Debug>:Debug>")
+    endif()
 endif()
 
-list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules)
+# project version
 include(FunctionReadVersion)
+read_version("${CMAKE_CURRENT_SOURCE_DIR}/version.pri" QBT_VER_MAJOR QBT_VER_MINOR QBT_VER_BUGFIX QBT_VER_BUILD QBT_VER_STATUS)
+message(STATUS "Project version is: ${QBT_VER_MAJOR}.${QBT_VER_MINOR}.${QBT_VER_BUGFIX}.${QBT_VER_BUILD} (${QBT_VER_STATUS})")
 
-read_version("${CMAKE_CURRENT_SOURCE_DIR}/version.pri" VER_MAJOR VER_MINOR VER_BUGFIX VER_BUILD VER_STATUS)
-# message(STATUS "Project version is: ${VER_MAJOR}.${VER_MINOR}.${VER_BUGFIX}.${VER_BUILD} (${VER_STATUS})")
-
-project(qBittorrent VERSION ${VER_MAJOR}.${VER_MINOR}.${VER_BUGFIX}.${VER_BUILD})
+project(qBittorrent VERSION ${QBT_VER_MAJOR}.${QBT_VER_MINOR}.${QBT_VER_BUGFIX}.${QBT_VER_BUILD})
 
 # check for invalid compiler version/CXX standard as early as possible
+# ideally, we would also be able to check whether or not libtorrent was compiled
+# with the same standard from imported target properties
 include(FunctionQbtCXXCompilerAndModeCheck)
 qbt_minimum_cxx_mode_check(14)
-message(STATUS "Building in C++${CMAKE_CXX_STANDARD} mode.\n"
-            "Make sure libtorrent was built with the same C++ mode for ABI compatibility.")
-
-set(PROJECT_VERSION "${VER_MAJOR}.${VER_MINOR}.${VER_BUGFIX}")
-
-if (NOT VER_BUILD EQUAL 0)
-    set(PROJECT_VERSION "${PROJECT_VERSION}.${VER_BUILD}")
-endif()
-
-set(PROJECT_VERSION "${PROJECT_VERSION}${VER_STATUS}")
-
-add_definitions(-DQBT_VERSION_MAJOR=${VER_MAJOR})
-add_definitions(-DQBT_VERSION_MINOR=${VER_MINOR})
-add_definitions(-DQBT_VERSION_BUGFIX=${VER_BUGFIX})
-add_definitions(-DQBT_VERSION_BUILD=${VER_BUILD})
-
-add_definitions(-DQBT_VERSION="v${PROJECT_VERSION}")
-add_definitions(-DQBT_VERSION_2="${PROJECT_VERSION}")
-
-include(GNUInstallDirs)
-include(FeatureSummary)
 
 # version requirements
-set(requiredBoostVersion 1.40)
+set(requiredBoostVersion 1.60)
 set(requiredQtVersion 5.9.0)
 set(requiredOpensslVersion 1.0)
-set(requiredLibtorrentVersion 1.1.10)
+set(requiredLibtorrentVersion 1.2.0)
+set(requiredZlibVersion 1.2.5.2)
 
-if (WIN32)
-    include(winconf)
+# features (some are platform-specific)
+include(FeatureSummary)
+include(FeatureOptionsSetup)
+feature_option(COUNTRIES_RESOLUTION "Enable resolving peers IP addresses to countries; default: ON" ON)
```

Comment:
Why did you drop `optional_compile_definitions` use? It adds corresponding compiler definition as well.

Reply:
@glassez 
Because I wanted to decouple the logic of adding compile definitions from the top-level CMake file. Now the compile definitions are added in the relevant places, depending on the feature. This makes it easier to migrate to a more target-based approach later across all components (e.g using `target_compile_definitions` instead of `add_compile_definitions`).